src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -248,23 +259,43 @@ public void Reset()
             _abortedCts = null;
         }
 
-        protected void ResetComponents()
+        protected void ResetComponents(bool immediateReuse)
         {
-            var frameHeaders = Interlocked.Exchange(ref _frameHeaders, null);
-            if (frameHeaders != null)
+            if (immediateReuse && HeaderFactory != null)
```

Comment:
I'm not in love with the `HeaderFactory != null` heuristic for caching headers. I would prefer this to be a first class option if it should configurable at all. Even if @davidfowl hates flags.